generator/common/beatgen/beatgen.go

```diff
@@ -0,0 +1,183 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package beatgen
+
+import (
+	"bufio"
+	"fmt"
+	"go/build"
+	"os"
+	"path/filepath"
+	"strings"
+
+	devtools "github.com/elastic/beats/dev-tools/mage"
+	"github.com/elastic/beats/generator/common/beatgen/setup"
+	"github.com/magefile/mage/mg"
+	"github.com/magefile/mage/sh"
+	"github.com/pkg/errors"
+)
+
+// ConfigItem represents a value that must be configured for the custom beat
+type ConfigItem struct {
+	Key     string
+	Default func(map[string]string) string
+}
+
+// required user config for a custom beat
+// These are specified in env variables with newbeat_*
+var configList = []ConfigItem{
+	{
+		Key: "project_name",
+		Default: func(cfg map[string]string) string {
+			return "examplebeat"
+		},
+	},
+	{
+		Key: "github_name",
+		Default: func(cfg map[string]string) string {
+			return "your-github-name"
+		},
+	},
+	{
+		Key: "beat_path",
+		Default: func(cfg map[string]string) string {
+			ghName, _ := cfg["github_name"]
+			beatName, _ := cfg["project_name"]
+			return "github.com/" + ghName + "/" + strings.ToLower(beatName)
+		},
+	},
+	{
+		Key: "full_name",
+		Default: func(cfg map[string]string) string {
+			return "Firstname Lastname"
+		},
+	},
+	{
+		Key: "type",
+		Default: func(cfg map[string]string) string {
+			return "beat"
+		},
+	},
+}
+
+// Generate generates a new custom beat
+func Generate() error {
+	cfg, err := getConfig()
+	if err != nil {
+		return err
```

Comment:
Error description example here:

Errors you can get here are if you track down the call stack:
* `bufio: buffer full`. It won't help users to understand what they have done wrong... actually even we can get confused about it. So I'm suggesting something like `error getting config: couldn't get config object from user: bufio: buffer full`. (it's just an example)